crm_helpdesk/tests/test_crm_helpdesk.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+# © 2017 Eficent Business and IT Consulting Services S.L. (www.eficent.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo.tests.common import TransactionCase
+from odoo.modules.module import get_module_resource
+from odoo.loglevels import ustr
+
+
+class TestCrmHelpdesk(TransactionCase):
+
+    def setUp(self):
+        super(TestCrmHelpdesk, self).setUp()
+
+        self.helpdesk_user = self.env['res.users'].create({
+            'company_id': self.env.ref("base.main_company").id,
+            'name': "Helpdesk User",
+            'login': "hdu",
+            'email': "helpdeskuser@yourcompany.com",
+            'groups_id': [(6, 0, [
+                self.env.ref('sales_team.group_sale_salesman').id,
+                self.env.ref('base.group_partner_manager').id])]
+        })
+
+    def test_crm_helpdesk_form(self):
+        helpdesk = self.env['crm.helpdesk'].sudo(self.helpdesk_user).create(
+            {'name': "Helpdesk 1"})
+        helpdesk.partner_id = self.env.ref('base.res_partner_1')
+
+    def test_crm_helpdesk_mail(self):
+        # Customer has Questions regarding our products, so he sent an email.
+        # Mail script will fetch his question from mail server.
+        # Then that mail is processed after reading EML file.
+        request_file = open(get_module_resource(
+            'crm_helpdesk', 'tests', 'customer_question.eml'), 'rb')
+        request_message = request_file.read()
+        self.env['crm.helpdesk'].sudo(self.helpdesk_user).\
+            message_process('crm.helpdesk', request_message)
+
+        # After getting the mail, the details of new question are checked.
+        question = self.env['crm.helpdesk'].\
+            sudo(self.helpdesk_user).search(
+            [('email_from', '=', 'Mr. John Right <info@customer.com>')],
+            limit=1)
+        self.assertTrue(question.ids,
+                        "Question is not created after getting request")
+        self.assertTrue(
+            question.name == ustr(
+                "Where is download link of user manual of your product ? "),
+            "Subject does not match")
+
+        # Now the message is updated according to provide services.
+        question = self.env['crm.helpdesk'].\
+            sudo(self.helpdesk_user).search(
+            [('email_from', '=', 'Mr. John Right <info@customer.com>')],
+            limit=1)
+        try:
+            self.env['crm.helpdesk'].\
+                sudo(self.helpdesk_user).message_update(
+                question,
+                {'subject': 'Link of product', 'body': 'www.odoo.com'})
+        except Exception:
+            pass
```

Comment:
What is the purpose of this try-except block?

Reply:
Don't know, it was from odoo v8 <code>help-desk.yml</code> file

Reply:
In any case it is incorrect, it comes from https://github.com/odoo/odoo/blob/8.0/addons/crm_helpdesk/test/process/help-desk.yml#L25. So it should be: `question.message_update({'subject': 'Link of product', 'body': 'www.odoo.com'})` in the new api. And then instead of `pass` I would `raise` to have the tests failing if something goes wrong.

Reply:
No. See https://github.com/OCA/OCB/blob/10.0/addons/mail/models/mail_thread.py#L1322

Reply:
look at the method you refer:

`def message_update(self, msg_dict, update_vals=None):`

you are calling it here like this:

`(...).message_update(question, {'subject': 'Link of product', 'body': 'www.odoo.com'})`

there, `question` is [the result of a search on helpdesk model](https://github.com/OCA/crm/pull/166/files#diff-905bd1d8172234ad0a77d084c9e66c30R53) and that dictionary is what is expected for `msg_dict`, not for `update_vals`

So, I recall, it should be:
`question.message_update({'subject': 'Link of product', 'body': 'www.odoo.com'})`

Reply:
Done